flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/internal/TransactionNamingStrategyImpl.java

```diff
@@ -53,6 +55,20 @@ public FlinkKafkaInternalProducer<byte[], byte[]> getTransactionalProducer(
             }
             return context.getProducer(context.buildTransactionalId(expectedCheckpointId));
         }
+    },
+    POOLING {
+        @Override
+        public FlinkKafkaInternalProducer<byte[], byte[]> getTransactionalProducer(
+                Context context) {
+            Collection<String> usedTransactionalIds = context.getOngoingTransactions();
+            for (int offset = 0; ; offset++) {
+                String transactionalIdCandidate = context.buildTransactionalId(offset);
+                if (usedTransactionalIds.contains(transactionalIdCandidate)) {
+                    continue;
+                }
+                return context.getProducer(transactionalIdCandidate);
```

Comment:
Do we have a log line somewhere that indicates how big the pool is? Since it might be an indicator of an serious issues we should probably log on every new producer creation.